flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,235 @@
+        getArchiveBaseName().convention(getProject().getName());
+        getArchiveExtension().convention("jar");
+        ObjectFactory objects = getProject().getObjects();
+        launcherClassName = objects.property(String.class).convention(Flask.Constants.DEFAULT_LAUNCHER_NAME);
+        mainClassName = objects.property(String.class);
+        jvmArgs = objects.listProperty(String.class).convention(new ArrayList<>());
+        javaAgents = new ArrayList<>();
+        from(getProject().tarTree(LauncherResource.instance), copySpec -> exclude(JarFile.MANIFEST_NAME));
+    }
+
+    @Input
+    @SneakyThrows
+    public String getLauncherArchiveHash() {
+        MessageDigest md5 = MessageDigest.getInstance("MD5");
+        try(InputStream stream = new DigestInputStream(LauncherResource.instance.read(), md5)) {
+            byte[] buffer = new byte[0x10000];
+            while(stream.read(buffer) >= 0) {}
+        }
+        return String.format("%032x", new BigInteger(1, md5.digest()));
+    }
+
+    @RequiredArgsConstructor
+    private static class StreamAction implements CopyActionProcessingStreamAction {
+
+        private final ZipOutputStream zoos;
+        private final Manifest manifest;
+        private final MessageDigest md;
+        private final byte[] buffer;
+
+        private final List<FileCopyDetailsInternal> fileCopyDetailsInternals = new ArrayList<>();
+
+        @SneakyThrows
+        public void write() {
+            ZipEntry zipEntry = new ZipEntry(Flask.Constants.METADATA_FOLDER + '/');
+            zoos.putNextEntry(zipEntry);
+            zipEntry = new ZipEntry(JarFile.MANIFEST_NAME);
+            zipEntry.setMethod(ZipEntry.DEFLATED);
+            zoos.putNextEntry(zipEntry);
+            manifest.write(zoos);
+            zoos.closeEntry();
+            for(FileCopyDetails fileCopyDetails : fileCopyDetailsInternals) {
+                String entryName = fileCopyDetails.getRelativePath().toString();
+                if(Objects.equals(Flask.Constants.METADATA_FOLDER, entryName)) continue;
+                zipEntry = new ZipEntry(entryName);
+                if(fileCopyDetails.isDirectory()) {
+                    zipEntry = new ZipEntry(entryName + '/');
+                    zoos.putNextEntry(zipEntry);
+                } else {
+                    boolean compressed = Flask.splitExtension(fileCopyDetails.getSourceName())
+                            .map(entry -> Objects.equals(".jar", entry.getValue()))
+                            .orElse(false);
+                    if (!compressed) {
+                        zipEntry.setMethod(ZipEntry.DEFLATED);
+                    } else {
+                        try (InputStream is = new FileInputStream(fileCopyDetails.getFile())) {
+                            Flask.computeSizeAndCrc32(zipEntry, is, buffer);
+                        }
+                        zipEntry.setMethod(ZipEntry.STORED);
+                    }
```

Comment:
I think this is a micro-optimisation. Jars tend to `STORE` the directory entries and `DEFLATE` the file entries regardless, which means that Java calculates the CRC32 for you automatically. If the flask contains almost entirely jars then you could always set the compression level to [NO_COMPRESSION](https://docs.oracle.com/javase/8/docs/api/java/util/zip/Deflater.html#NO_COMPRESSION) instead.

Reply:
it is micro-optimization but there was also another issue using the `Jar` task: I need to pre-configure the task to include the manifest entries I want it to include, the `LIB-INF` folder and so on but I also want to let the user to run its own task configuration (possibly even multiple times). 

Also, my configuration has to run after the user configuration since the manifest entries (like the main class name) needs to be provided by the user.

 The way I did it before, overriding the `configure` method appending a callback that was adding the LIB-INF folder and so on, every time the task configuration ran all the entries were added (with a resulting archive with duplicated entries)

[this](https://github.com/woggioni/corda-gradle-plugins/blob/68ecde29863964c946f365a1cf288a585dbdd2b1/flask/src/main/groovy/net/corda/gradle/flask/FlaskPlugin.groovy#L118-L138) was the broken code